File: bindings/go/examples/multi_agent.go
// Multi-agent coordination: multiple agents collaborating in a shared context.
package main

import (
	"fmt"
	"log"
	"sync"

	scp "github.com/limn/scp-go"
)

func runAgent(name string, identity *scp.Identity, contextID string, wg *sync.WaitGroup) {
	defer wg.Done()

	ctx, err := scp.JoinContext(identity, contextID)
	if err != nil {
		log.Printf("[%s] Failed to join: %v", name, err)
		return
	}
	fmt.Printf("[%s] Joined context %s\n", name, contextID)

	if err := ctx.Send([]byte(fmt.Sprintf("[%s] reporting in", name))); err != nil {
		log.Printf("[%s] Send failed: %v", name, err)
		return
	}

	count := 0
	for msg := range ctx.Receive() {
		sender := msg.SenderDID
		if len(sender) > 16 {
			sender = sender[:16]
		}
		fmt.Printf("[%s] Received from %s...: %s\n", name, sender, msg.Content)
		count++
		if count >= 2 {
			break
		}
	}

	ctx.Leave()
	fmt.Printf("[%s] Left context\n", name)
}

func main() {
	if err := scp.Init(); err != nil {
		log.Fatal(err)
	}
	defer scp.Shutdown()

	// Create identities for coordinator and two agents
	coordinator, err := scp.NewIdentity("platform")
	if err != nil {
		log.Fatal(err)
	}
	agentA, err := scp.NewIdentity("platform")
	if err != nil {
		log.Fatal(err)
	}
	agentB, err := scp.NewIdentity("platform")
	if err != nil {
		log.Fatal(err)
	}

	// Coordinator creates the context
	ctx, err := scp.NewContext(coordinator, scp.ContextParams{
		Ceiling:    []string{"msg:send", "msg:receive", "tool:invoke"},
		Governance: "single_admin",
		Roles: map[string][]string{
			"agent": {"msg:send", "msg:receive", "tool:invoke"},
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	defer ctx.Close()
	fmt.Println("Context created:", ctx.ContextID())

	// Mint UCANs for each agent
	_, err = scp.MintUcan(coordinator, agentA.DID(), []string{"msg:send", "msg:receive"}, ctx.ContextID())
	if err != nil {
		log.Fatal(err)
	}
	_, err = scp.MintUcan(coordinator, agentB.DID(), []string{"msg:send", "msg:receive"}, ctx.ContextID())
	if err != nil {
		log.Fatal(err)
	}

	// Run agents concurrently
	var wg sync.WaitGroup
	wg.Add(2)
	go runAgent("Agent-A", agentA, ctx.ContextID(), &wg)
	go runAgent("Agent-B", agentB, ctx.ContextID(), &wg)
	wg.Wait()
}

File: bindings/kotlin/examples/MultiAgent.kt
/**
 * Multi-agent coordination: multiple agents collaborating in a shared context.
 *
 * Demonstrates identity creation, context creation with governance, UCAN
 * minting, message sending, and message receiving using the actual Kotlin
 * SDK API surface via CoroutineBridge.
 *
 * Prerequisites:
 *   implementation("works.limn:scp-kt:0.1.0")
 *
 * Usage:
 *   ./gradlew run --args="multi-agent"
 */

package works.limn.scp.examples

import kotlinx.coroutines.async
import kotlinx.coroutines.flow.take
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.putJsonArray
import works.limn.scp.CustodyType
import works.limn.scp.bridge.CoroutineBridge

suspend fun runAgent(
    name: String,
    agentHandle: Long,
    bridge: CoroutineBridge,
    contextHandle: Long,
) {
    // Agent joins the existing context
    bridge.context.join(contextHandle, agentHandle)
    println("[$name] Joined context")

    // Send a message
    bridge.context.send(contextHandle, agentHandle, "[$name] reporting in".toByteArray())
    println("[$name] Sent message")

    // Subscribe to messages via Flow
    val subscription = bridge.context.subscribe(contextHandle)
    var count = 0
    subscription.take(2).collect { messageJson ->
        println("[$name] Received: $messageJson")
        count++
    }

    bridge.context.leave(contextHandle, agentHandle)
    println("[$name] Left context")
}

fun multiAgentExample(bridge: CoroutineBridge) = runBlocking {
    // Create identities for coordinator and two agents
    val coordinatorHandle = bridge.identity.create(CustodyType.IN_MEMORY)
    val agentAHandle = bridge.identity.create(CustodyType.IN_MEMORY)
    val agentBHandle = bridge.identity.create(CustodyType.IN_MEMORY)
    println("Coordinator: $coordinatorHandle")
    println("Agent A: $agentAHandle")
    println("Agent B: $agentBHandle")

    // Coordinator creates the context with governance capabilities
    val paramsJson = buildJsonObject {
        putJsonArray("ceiling") {
            add(JsonPrimitive("messages:read"))
            add(JsonPrimitive("messages:write"))
            add(JsonPrimitive("outlet:call:*"))
            add(JsonPrimitive("member:invite"))
            add(JsonPrimitive("member:remove"))
            add(JsonPrimitive("role:assign"))
        }
        put("governance", JsonPrimitive("single_admin"))
        put("memory_scope", JsonPrimitive("ephemeral"))
    }.toString()

    val contextHandle = bridge.context.create(coordinatorHandle, paramsJson)
    println("Context created: $contextHandle")

    // Mint UCANs for each agent via the bridge
    //   bridge.ucan.mint(contextHandle, agentADid, """["messages:write","messages:read"]""")
    //   bridge.ucan.mint(contextHandle, agentBDid, """["messages:write","messages:read"]""")
    // (requires agent DIDs -- omitted in this example since handles are opaque Longs)

    // Run agents concurrently
    val taskA = async { runAgent("Agent-A", agentAHandle, bridge, contextHandle) }
    val taskB = async { runAgent("Agent-B", agentBHandle, bridge, contextHandle) }
    taskA.await()
    taskB.await()

    bridge.context.close(contextHandle, coordinatorHandle)
    println("Context closed.")
}
